kubernetes/cache.go

```diff
@@ -0,0 +1,325 @@
+		}
+		nsPods := make([]v1.Pod, len(pods))
+		for i, pod := range pods {
+			nsPods[i] = *(pod.(*v1.Pod))
+		}
+		return nsPods, true
+	}
+	return nil, false
+}
+
+func (c *controllerImpl) GetReplicationControllers(namespace string) ([]v1.ReplicationController, bool) {
+	indexer := c.controllers["ReplicationController"].GetIndexer()
+	repcons, err := indexer.ByIndex("namespace", namespace)
+	if err != nil {
+		return nil, false
+	}
+	if len(repcons) > 0 {
+		_, ok := repcons[0].(*v1.ReplicationController)
+		if !ok {
+			return nil, false
+		}
+		nsRepcons := make([]v1.ReplicationController, len(repcons))
+		for i, repcon := range repcons {
+			nsRepcons[i] = *(repcon.(*v1.ReplicationController))
+		}
+		return nsRepcons, true
+	}
+	return nil, false
+}
+
+func (c *controllerImpl) GetReplicaSets(namespace string) ([]v1beta2.ReplicaSet, bool) {
+	indexer := c.controllers["ReplicaSet"].GetIndexer()
+	repsets, err := indexer.ByIndex("namespace", namespace)
+	if err != nil {
+		return nil, false
+	}
+	if len(repsets) > 0 {
+		_, ok := repsets[0].(*v1beta2.ReplicaSet)
+		if !ok {
+			return nil, false
+		}
+		nsRepsets := make([]v1beta2.ReplicaSet, len(repsets))
+		for i, repset := range repsets {
+			nsRepsets[i] = *(repset.(*v1beta2.ReplicaSet))
+		}
+		return nsRepsets, true
+	}
+	return nil, false
+}
+
+func (c *controllerImpl) GetStatefulSet(namespace, name string) (*v1beta2.StatefulSet, bool) {
+	indexer := c.controllers["StatefulSet"].GetIndexer()
+	fulsets, exist, err := indexer.GetByKey(namespace + "/" + name)
+	if err != nil {
+		return nil, false
+	}
+	if exist {
+		fulset, ok := fulsets.(*v1beta2.StatefulSet)
+		if !ok {
+			return nil, false
+		}
+		return fulset, true
+	}
+	return nil, false
+}
+
+func (c *controllerImpl) GetStatefulSets(namespace string) ([]v1beta2.StatefulSet, bool) {
+	indexer := c.controllers["StatefulSet"].GetIndexer()
+	fulsets, err := indexer.ByIndex("namespace", namespace)
+	if err != nil {
+		return nil, false
+	}
+	if len(fulsets) > 0 {
+		_, ok := fulsets[0].(*v1beta2.StatefulSet)
+		if !ok {
+			return nil, false
+		}
+		nsFulsets := make([]v1beta2.StatefulSet, len(fulsets))
+		for i, fulset := range fulsets {
+			nsFulsets[i] = *(fulset.(*v1beta2.StatefulSet))
+		}
+		return nsFulsets, true
+	}
+	return nil, false
+}
+
+func (c *controllerImpl) GetService(namespace, name string) (*v1.Service, bool) {
+	indexer := c.controllers["Service"].GetIndexer()
+	services, exist, err := indexer.GetByKey(namespace + "/" + name)
+	if err != nil {
+		return nil, false
+	}
+	if exist {
+		service, ok := services.(*v1.Service)
+		if !ok {
+			return nil, false
+		}
+		return service, true
+	}
+	return nil, false
+}
+
+func (c *controllerImpl) GetServices(namespace string) ([]v1.Service, bool) {
+	indexer := c.controllers["Service"].GetIndexer()
+	services, err := indexer.ByIndex("namespace", namespace)
+	if err != nil {
+		return nil, false
+	}
+	if len(services) > 0 {
+		_, ok := services[0].(*v1.Service)
+		if !ok {
```

Comment:
If I understand correctly, this is to check that the cache contains objects of the expected type?
So, if not, maybe we should propagate some kind of "Internal Error", no? (and log)

Reply:
+1